### DISCUSSION TITLE
Dynamic URL

### DISCUSSION THREAD

**Discussion by readonlymem0ry** [pschatzmann/arduino-audio-tools#1535] [Q&A] [answered]
How to make dynamic url audio source? all examples is using constant variable and run it in setup() with url.begin() function, correct me if i'm wrong. But, the point is how to use dynamic url in the sketch because i want to change the url based on API data and how to run url.begin() on loop(), it is possible or not? if it possible please teach me. Sorry for my bad English.

**Comment by pschatzmann** [comment, marked as answer]
Did you look at the class documentation ? 

There is nothing which would prevent to call begin just whenever if make sense. Just don't forget to call end() if you already have called begin() before.

**Reply by readonlymem0ry** [reply]
yes i did, but i don't understand how to use that class in the sketch, where do i should place the function? setup or loop?

**Reply by pschatzmann** [reply]
Wherever you need it!

**Reply by pschatzmann** [reply]
I don't understand: I don't see any logic to change the URL ?

**Reply by readonlymem0ry** [reply]
so can you give example?

**Comment by pschatzmann** [comment]
Not without knowing what you want to do!

**Reply by readonlymem0ry** [reply]
i want to make the url dinamically because the url is set from my API, just that, thanks for your patience, sorry

**Comment by pschatzmann** [comment]
You need to work on your communication skills ! Your API does not mean anything to me. Is it something that already exists ? What is the functionality, what is the data representation ?

You need to decide how and where you want to store your URLs. It could be in the limited RAM or it could be in PROGMEM. 

I suggest that you have a look at the AudioPlayer with the corresponding [AudioSourceURL](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_source_u_r_l.html) that uses an array of C strings in the constructor which are stored in PROGMEM when you define them as const.

Here is the related [example for an AudioKit](https://github.com/pschatzmann/arduino-audio-tools/blob/90ec7397af6d1ea2fee7ee311c77313913c5057e/examples/examples-communication/http-client/player-url_icy-audiokit/player-url_icy-audiokit.ino#L24)
